bot/common/tasks/tasks.py

```diff
@@ -0,0 +1,176 @@
+        last_sent = await cache.get(cache_key)
+        # check that today is the same weekday as the cadence and that
+        # the last update was sent 1 week ago
+        now: datetime = datetime.now()
+
+        # last_entry does not exist -> return delta to closest ocurrence of next run
+        if last_sent is None:
+            return self.get_timedelta(now)
+
+        last_sent: datetime = datetime.strptime(last_sent, DATETIME_CACHE_FMT)
+
+        # if the last_sent is on a day that does not match specified weekday,
+        # the cadence may have been changed in code, or the task ran late
+        # for some reason. log appropriately, return timedelta for next ocurrence
+        if last_sent.weekday() != self.day_to_run:
+            logger.warning(
+                f"Weekly task {cache_key} was last run on a different day "
+                f"({last_sent.strftime(DATETIME_CACHE_FMT)}) than is specified on "
+                f"the task ({self.time_to_run.isoformat()} on {self.day_to_run}th "
+                "day of the week)"
+            )
+            return self.get_timedelta(now)
+
+        return self.get_timedelta_with_last_sent(now, last_sent)
+
+    def get_timedelta(self, now: datetime) -> timedelta:
+        if now.weekday() == self.day_to_run:
+            if now.time() >= self.time_to_run:
+                # run immediately
+                return timedelta(seconds=-1)
+        # wait until the next ocurrence of the specified date and time
+        return self.get_timedelta_until_next_occurrence(now)
+
+    def get_timedelta_until_next_occurrence(self, now: datetime) -> timedelta:
+        days_until = (self.day_to_run - now.weekday()) % 7
+        next_ocurrence = now + timedelta(days=days_until)
+        next_ocurrence = datetime.combine(next_ocurrence.date(), self.time_to_run)
+        td = next_ocurrence - now
+        if td.total_seconds() <= 0:
+            td = td + timedelta(days=7)
+        return td
+
+    def get_timedelta_with_last_sent(
+        self, now: datetime, last_sent: datetime
+    ) -> timedelta:
+        # If the last_sent datetime is today, wait until next ocurrence
+        if now.date() == last_sent.date():
+            return self.get_timedelta_until_next_occurrence(now)
+        return self.get_timedelta(now)
+
+
+# conforms with date.weekday()
+class Days:
+    MONDAY = 0
+    TUESDAY = 1
+    WEDNESDAY = 2
+    THURSDAY = 3
+    FRIDAY = 4
+    SATURDAY = 5
+    SUNDAY = 6
+
+
+# TASKS
+tasks = TaskBatch(RedisCache(), TASK_LOOP_INTERVAL_MINUTES * 60)
+
+
+def task(task_collection: TaskBatch, name: str, cadence: Cadence):
+    logger.info(f"adding {name} to task list")
+
+    def inner(f):
+        task_collection.add_task(f, name, cadence)
+
+    return inner
+
+
+@task(tasks, "weekly_contribution_report", cadence=Weekly(Days.FRIDAY, time(15, 00)))
+async def weekly_contribution_report():
+    await send_weekly_contribution_reports()
```

Comment:
@alexkeating this will likely need to change. I'll likely inject a reference to the discord bot, but need to find a way to share the event loop used internally by the discord bot with the second thread used to iterate the tasks.

Reply:
Ideally, we have some event we can listen to with the bot that triggers the report. It might be worth looking if pycord allows a custom event. That is probably cleaner than sharing state across event loops

Reply:
agree, would prefer that approach to threading. I hadn't checked to see if events outside of reactions/messages/commands were supported. but just double checking that we would want kmalone to send this outside of any actual discord interaction.